src/cpu/intel.rs

```diff
@@ -35,139 +36,190 @@ mod abi_assumptions {
-                // is safe (no risk of illegal instructions) and correct (the
-                // value computed will be the same regardless).
-                let flag = unsafe { &avx2_available };
-                flag.store(1, Ordering::Relaxed);
-            }
-        }
+            super::cpuid_to_caps_and_set_c_flags(&cpuid)
+        });
 
-        cpu
+        // SAFETY: We initialized the CPU features as required.
+        // `INIT.call_once` has `happens-before` semantics.
+        unsafe { cpu::Features::new_after_feature_flags_written_and_synced_unchecked() }
     }
 
-    pub(super) fn get(_cpu_features: cpu::Features) -> &'static [u32; 4] {
-        prefixed_extern! {
-            static mut OPENSSL_ia32cap_P: [u32; 4];
-        }
-        // TODO(MSRV 1.82.0): Remove `unsafe`.
-        #[allow(unused_unsafe)]
-        let p = unsafe { ptr::addr_of!(OPENSSL_ia32cap_P) };
+    pub(super) fn get(_cpu_features: cpu::Features) -> u32 {
         // SAFETY: Since only `get_or_init()` could have created
-        // `_cpu_features`, and it only does so after the `INIT.call_once()`
-        // (which guarantees `happens-before` semantics), or during the
-        // `INIT.call_once()` after writing to `OPENSSL_ia32cap_P`, we can
-        // safely read from `OPENSSL_ia32cap_P` without further
-        // synchronization.
-        unsafe { &*p }
+        // `_cpu_features`, and it only does so after `FEATURES.call_once()`,
+        // we have met the prerequisites for calling `get_unchecked()`.
+        let features: &u32 = unsafe { FEATURES.get_unchecked() };
+        *features
     }
+
+    static FEATURES: spin::Once<u32> = spin::Once::new();
 }
 
-impl Feature {
-    #[allow(clippy::needless_return)]
-    #[inline(always)]
-    pub fn available(&self, cpu_features: super::Features) -> bool {
-        let flags = featureflags::get(cpu_features);
-        self.mask == self.mask & flags[self.word]
+fn cpuid_to_caps_and_set_c_flags(cpuid: &[u32; 4]) -> u32 {
+    // The `prefixed_extern!` uses below assume this
+    const _ATOMIC32_ALIGNMENT_EQUSLS_U32_ALIGNMENT: () =
+        assert!(align_of::<AtomicU32>() == align_of::<u32>());
+
+    fn check(cpuid: &[u32; 4], idx: usize, bit: u32) -> bool {
+        let shifted = 1 << bit;
+        (cpuid[idx] & shifted) == shifted
+    }
+    fn set(out: &mut u32, shift: Shift) {
+        let shifted = 1 << (shift as u32);
+        debug_assert_eq!(*out & shifted, 0);
+        *out |= shifted;
+        debug_assert_eq!(*out & shifted, shifted);
     }
-}
 
-macro_rules! impl_get_feature_cpuid {
-    { ( $( $arch:expr ),+ ) => ($word:expr, $bit:expr) => $NAME:ident => $Name:ident } => {
-        #[cfg(any( $( target_arch = $arch ),+ ))]
-        const $NAME: Feature = Feature {
-            word: $word,
-            mask: 1 << $bit,
-        };
+    let mut caps = 0;
 
-        #[cfg(any( $( target_arch = $arch ),+ ))]
-        impl_get_feature! { $NAME => $Name }
+    if check(cpuid, 0, 24) {
```

Comment:
It is expected, presently, that this condition is always true because we only support platforms that have SSE2 enabled by default, and there are not SSE2-enabled CPUs without FXSR.